refactor(*): make typedefs for all options params

iShibi opened the pull request:
**Please describe the changes this PR makes and why it should be merged:**

This PR starts the work of creating typedefs for `options` params of all the methods that don't have one. I've created a few so far but there are still few more to add. Since, #5784 added some new typedefs too, I'll rebase this PR once that one is merged to remove duplicate changes. This PR also tries to add more helpful description of a property in places where the old one is a little vague.

ℹ️ The names for the new typedefs are taken from their respective interfaces in the typings.

**Status and versioning classification:**
- This PR **only** includes non-code changes, like changes to documentation, README, etc.
<!--
Please move lines that apply to you out of the comment:
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
-->

ImRodry commented on src/managers/GuildChannelManager.js:
```diff
@@ -59,21 +59,26 @@ class GuildChannelManager extends BaseManager {
+  /**
+   * Creates a new channel in the guild.
+   * @param {string} name The name of the new channel
+   * @param {GuildCreateChannelOptions} [options={}] Options for creating the new channel
```
I think having [options = {}] is redundant because the [] indicates that the parameter is optional and saying that an empty object is assumed is already implicit when the parameter is omitted. I know this is done many times across the lib but I still think it's worth the change (maybe change others too, would need feedback from the maintainers)

iShibi replied:
Yes, the `[]` states that the param is optional. Doing `[options = {}]` shows `{}` under default in the docs which will not be there if we remove the `= {}` part. I think this is better because it distinct itself from optional param that are not objects.

ImRodry replied:
Yeah I know how that's displayed but if the options parameter is already marked as an object (by the typedef), adding {} wouldn't add much information imo

ImRodry commented on src/managers/GuildManager.js:
```diff
@@ -128,24 +128,29 @@ class GuildManager extends BaseManager {
+   * the first element of this array is used to change properties of the guild's everyone role.
+   * @param {number} [systemChannelID] The ID of the system channel
+   * @param {SystemChannelFlagsResolvable} [systemChannelFlags] The flags of the system channel
+   * @param {VerificationLevel} [verificationLevel] The verification level for the guild
```
typedefs have `property`'s, not `param`'s

iShibi replied:
This is what happens when you work at the time when you should be sleeping, will fix soon 👍

ImRodry replied:
xD all good

ImRodry reviewed (commented):
I know this is still a work in progress but I'm interested to see progress on this, also here are some changes I think should be made.

ImRodry commented on src/structures/GuildChannel.js:
```diff
@@ -492,19 +502,24 @@ class GuildChannel extends Channel {
-   * Creates an invite to this guild channel.
-   * @param {Object} [options={}] Options for the invite
-   * @param {boolean} [options.temporary=false] Whether members that joined via the invite should be automatically
+   * Options used to create an invite to a guild.
```
```suggestion
   * Options used to create an invite to a guild channel.
```

ImRodry commented on src/structures/GuildChannel.js:
```diff
@@ -554,22 +569,15 @@ class GuildChannel extends Channel {
-  /* eslint-disable max-len */
+  /**
+   * Options used to clone a guild channel.
+   * @typedef {GuildCreateChannelOptions} GuildChannelCloneOptions
```
Is it clone or create? And if it’s clone, where did the other properties go?

iShibi replied:
Clone. Those are inherited from the `GuildCreateChannelOptions` typedef.

ImRodry replied:
But this is that typedef, am I missing something?

iShibi replied:
This is `GuildChannelCloneOptions` typedef which inherits props from the typedef that is inside the `{}` on the left and is defined [here](https://github.com/discordjs/discord.js/blob/9dd3a9a018e0c13ba448f6e579bc866d68c75dc1/src/managers/GuildChannelManager.js#L63)

ImRodry replied:
Ohhh yeah right my bad lol

ImRodry commented on src/structures/MessageMentions.js:
```diff
@@ -161,14 +161,19 @@ class MessageMentions {
   }
 
+  /**
+   * Options used to check a mention.
```
```suggestion
   * Options used to check for a mention.
```

ImRodry commented on typings/index.d.ts:
```diff
@@ -1049,7 +1049,7 @@ declare module 'discord.js' {
 
   // tslint:disable-next-line:no-empty-interface - Merge RateLimitData into RateLimitError to not have to type it again
-  interface RateLimitError extends RateLimitData { }
+  interface RateLimitError extends RateLimitData {}
```
Watch out for this change (and others caused by vscode's default formatter) but why don't you address the comment while you're at it?

iShibi replied:
The linter fixes all the formating during commit. That comment isn't mine.

ImRodry replied:
I know the comment isn’t yours but you could address it I guess
But those formatting issues are caused by vscode’s default formatter, just use prettier

ImRodry replied:
Seems like they were fixed in a later commit, nvm

iShibi replied:
The current formating is the correct one as it is same in the master. You commented on it just after rebase when this change was introduced. Since, no commit happened during rebase the linter couldn't have fixed it. It fixed it in the later commits, that's why this suggestion is showing as outdated.

I don't know what you want me to address about the comment. Could you be a little more specific?

ImRodry replied:
Ah alright my bad then
Since the comment says “Merge RateLimitData into RateLimitError to not have to type it again” you could address that since you’re touching interfaces and typings, it’d be in line with the theme of the PR imo

iShibi commented:
I think I've done all the neccessary changes, so I'll mark this as ready for review. This PR along with #5805 should make the docs and typings a little bit more clean 🥳

kyranet commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * @property {ExplicitContentFilterLevel} [explicitContentFilter] The explicit content filter level for the guild
+   * @property {BufferResolvable|Base64Resolvable} [icon=null] The icon for the guild
+   * @property {string} [region] The region for the server, defaults to the closest one available
+   * @property {PartialRoleData[]} [roles] The roles for this guild,
```
```suggestion
   * @property {PartialRoleData[]} [roles=[]] The roles for this guild,
```

kyranet commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * @typedef {Object} GuildCreateOptions
+   * @property {number} [afkChannelID] The ID of the AFK channel
+   * @property {number} [afkTimeout] The AFK timeout in seconds
+   * @property {PartialChannelData[]} [channels] The channels for this guild
```
```suggestion
   * @property {PartialChannelData[]} [channels=[]] The channels for this guild
```

kyranet commented on src/structures/MessageMentions.js:
```diff
@@ -161,14 +161,19 @@ class MessageMentions {
+  /**
+   * Options used to check for a mention.
+   * @typedef {Object} MessageMentionsHasOptions
+   * @property {boolean} [ignoreDirect=false] - Whether to ignore direct mentions to the item
+   * @property {boolean} [ignoreRoles=false] - Whether to ignore role mentions to a guild member
+   * @property {boolean} [ignoreEveryone=false] - Whether to ignore everyone/here mentions
```
```suggestion
   * @property {boolean} [ignoreDirect=false] Whether to ignore direct mentions to the item
   * @property {boolean} [ignoreRoles=false] Whether to ignore role mentions to a guild member
   * @property {boolean} [ignoreEveryone=false] Whether to ignore everyone/here mentions
```

kyranet reviewed (changes requested):
I'm unsure if jsdocs forbid `[field=[]]`, but I figured I'd point out just in case.

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+  /**
+   * Options used to create a guild.
+   * @typedef {Object} GuildCreateOptions
+   * @property {number} [afkChannelID] The ID of the AFK channel
```
This can be a Snowflake too

iShibi replied:
hmm, now that you pointed it out, I don't get why it is a number in the first place. Should be changed to `Snowflake` in typings too I think.

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * @typedef {Object} GuildCreateOptions
+   * @property {number} [afkChannelID] The ID of the AFK channel
+   * @property {number} [afkTimeout] The AFK timeout in seconds
+   * @property {PartialChannelData[]} [channels=[]] The channels for this guild
```
Just so I'm sure, is this defined?

iShibi replied:
Yup, it's defined [here](https://github.com/discordjs/discord.js/blob/da6f21a3fca084f27d115754fed5fff8091cce77/src/managers/GuildManager.js#L74). Same, issue there tho, the id props are documented as `number`.

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * for the guild
+   * @property {ExplicitContentFilterLevel} [explicitContentFilter] The explicit content filter level for the guild
+   * @property {BufferResolvable|Base64Resolvable} [icon=null] The icon for the guild
+   * @property {string} [region] The region for the server, defaults to the closest one available
```
n/a anymore

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * @property {ExplicitContentFilterLevel} [explicitContentFilter] The explicit content filter level for the guild
+   * @property {BufferResolvable|Base64Resolvable} [icon=null] The icon for the guild
+   * @property {string} [region] The region for the server, defaults to the closest one available
+   * @property {PartialRoleData[]} [roles=[]] The roles for this guild,
```
As PartialChannelData, just to be sure, is this defined?

iShibi replied:
Yup, [here](https://github.com/discordjs/discord.js/blob/da6f21a3fca084f27d115754fed5fff8091cce77/src/managers/GuildManager.js#L52)

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -128,23 +128,29 @@ class GuildManager extends BaseManager {
+   * @property {string} [region] The region for the server, defaults to the closest one available
+   * @property {PartialRoleData[]} [roles=[]] The roles for this guild,
+   * the first element of this array is used to change properties of the guild's everyone role.
+   * @property {number} [systemChannelID] The ID of the system channel
```
Can be a snowflake too

iShibi replied:
Same as above, how come no one noticed it for so long 😆

vladfrangu commented on src/managers/GuildMemberManager.js:
```diff
@@ -198,19 +198,19 @@ class GuildMemberManager extends BaseManager {
-   * @property {boolean} [dry] Get number of users that will be kicked, without actually kicking them
-   * @property {boolean} [count] Whether or not to return the number of users that have been kicked.
+   * @property {number} [days=7] Number of days of inactivity required to kick
+   * @property {boolean} [dry=false] Get number of users that will be kicked, without actually kicking them
```
```suggestion
   * @property {boolean} [dry=false] Get the number of users that will be kicked, without actually kicking them
```

vladfrangu commented on src/rest/RequestHandler.js:
```diff
@@ -125,13 +125,7 @@ class RequestHandler {
-         * @param {string} rateLimitInfo.path Path used for request that triggered this event
-         * @param {string} rateLimitInfo.route Route used for request that triggered this event
-         * @param {boolean} rateLimitInfo.global Whether the rate limit that was reached was the global limit
+         * @param {RateLimitData} rateLimitData Object containing the rate limit info
```
Make sure these have the same properties 🙏

iShibi replied:
yup. [RateLimitData](https://github.com/discordjs/discord.js/blob/da6f21a3fca084f27d115754fed5fff8091cce77/src/util/Constants.js#L8) defined in Constants, has exactly the same props.

vladfrangu commented on src/rest/RequestHandler.js:
```diff
@@ -231,12 +225,17 @@ class RequestHandler {
+         */
+
         /**
          * Emitted periodically when the process sends invalid messages to let users avoid the
```
```suggestion
         * Emitted periodically when the process sends invalid requests to let users avoid the
```

iShibi commented:
Changed, `number` to `Snowflake` for IDs. Also, `PartialChannelData#permissionOverwrites` is an array but wasn't documented as one, fixed that too. Added `PartialOverwriteData` interface for the same in typings.

iCrawl commented:
This needs a rebase.

iShibi commented:
Rebased

vladfrangu commented on src/managers/GuildManager.js:
```diff
@@ -50,7 +50,7 @@ class GuildManager extends BaseManager {
    * Partial data for a Role.
    * @typedef {Object} PartialRoleData
-   * @property {number} [id] The ID for this role, used to set channel overrides,
+   * @property {Snowflake} [id] The ID for this role, used to set channel overrides,
```
I said this can also be a number for a reason! You can specify a role as `{ id: 1, color: 0xffffff }` or `{ id: "8000000000000000000", role: 0x000000 }`. The snowflake type is ``type Snowflake = `${bigint}`;``, which is a string that can be a bigint

Ditto for the rest

iShibi replied:
You said, "it can be a `Snowflake` too", so I thought `number` is wrong type and changed it to `Snowflake` only, which is what we have as a type for most IDs. Fixed it in the latest commit.

SpaceEEC commented on src/managers/GuildChannelManager.js:
```diff
@@ -60,21 +60,26 @@ class GuildChannelManager extends BaseManager {
-   * @param {Object} [options] Options
-   * @param {string} [options.type='text'] The type of the new channel, either `text`, `voice`, `category`, `news`,
+   * Options used to create a new channel in a guild.
+   * @typedef {Object} GuildCreateChannelOptions
```
Maybe name this `GuildChannelCreateOptions` similar to the emoji one below?

iShibi replied:
Done

SpaceEEC commented on src/structures/GuildChannel.js:
```diff
@@ -529,19 +539,24 @@ class GuildChannel extends Channel {
-   * @param {Object} [options={}] Options for the invite
-   * @param {boolean} [options.temporary=false] Whether members that joined via the invite should be automatically
+   * Options used to create an invite to a guild channel.
+   * @typedef {Object} InviteOptions
```
Maybe name this `CreateInviteOptions`?
